Make completion work if the command line already has some accented characters.

diff -Naurdp mc-4.6.1.orig/src/complete.c mc-4.6.1/src/complete.c
--- mc-4.6.1.orig/src/complete.c	2005-07-23 18:52:02.000000000 +0200
+++ mc-4.6.1/src/complete.c	2007-01-02 17:37:53.000000000 +0100
@@ -797,7 +797,7 @@ static int insert_text (WInput *in, char
 	    *p = 0;
 	}
 	memcpy (in->buffer + start, text, len - start + end);
-	in->point += len;
+	in->point += mbstrnlen(text - start + end, len);
 	update_input (in, 1);
 	end += len;
     }
@@ -912,7 +912,7 @@ complete_engine (WInput *in, int what_to
     if (in->completions && in->point != end)
     	free_completions (in);
     if (!in->completions){
-    	end = in->point;
+    	end = columns_to_bytes(in->buffer, in->point);
         for (start = end ? end - 1 : 0; start > -1; start--)
     	    if (strchr (" \t;|<>", in->buffer [start]))
     	        break;
diff -Naurdp mc-4.6.1.orig/src/util.c mc-4.6.1/src/util.c
--- mc-4.6.1.orig/src/util.c	2007-01-02 16:08:29.000000000 +0100
+++ mc-4.6.1/src/util.c	2007-01-02 17:37:35.000000000 +0100
@@ -139,6 +139,34 @@ mbstrlen (const char *str)
 	return strlen (str);
 }
 
+size_t
+mbstrnlen (const char *str, size_t maxlen)
+{
+#ifdef UTF8
+    if (1) {
+        size_t width = 0;
+
+        for (; *str && maxlen; str++, maxlen--) {
+            wchar_t c;
+            size_t len;
+
+            len = mbrtowc (&c, str, maxlen > MB_CUR_MAX ? MB_CUR_MAX : maxlen, NULL);
+	    
+            if (len == (size_t)(-1) || len == (size_t)(-2)) break;
+	    
+            if (len > 0) {
+                int wcsize = wcwidth(c);
+                width += wcsize >= 0 ? wcsize : 1;
+                str += len-1;
+            }
+        }
+
+        return width;
+    } else
+#endif
+        return strnlen (str, maxlen);
+}
+
 int
 columns_to_bytes (const char *str, int col)
 {
diff -Naurdp mc-4.6.1.orig/src/util.h mc-4.6.1/src/util.h
--- mc-4.6.1.orig/src/util.h	2007-01-02 16:08:02.000000000 +0100
+++ mc-4.6.1/src/util.h	2007-01-02 16:58:50.000000000 +0100
@@ -95,6 +95,7 @@ char *get_owner (int);
 
 void fix_utf8(char *str);
 size_t mbstrlen (const char *);
+size_t mbstrnlen (const char *, size_t);
 int columns_to_bytes (const char *, int);
 wchar_t *mbstr_to_wchar (const char *);
 char *wchar_to_mbstr (const wchar_t *);
